core/crossmodel/applicationurl.go

```diff
@@ -43,146 +53,104 @@ func (u *ApplicationURL) Path() string {
-// based on the specified URL.
-func ApplicationDirectoryForURL(urlStr string) (string, error) {
-	url, err := ParseApplicationURL(urlStr)
-	if err != nil {
-		return "", err
-	}
-	return url.Directory, nil
+	url := ApplicationURL(*urlParts)
+	return &url, nil
 }
 
 // ApplicationURLParts contains various attributes of a URL.
 type ApplicationURLParts ApplicationURL
 
 // ParseApplicationURLParts parses a partial URL, filling out what parts are supplied.
-// TODO(wallyworld) update ParseApplicationURL to use this method and perform additional validation on top.
+// This method is used to generate a filter used to query matching application URLs.
 func ParseApplicationURLParts(urlStr string) (*ApplicationURLParts, error) {
-	url, err := gourl.Parse(urlStr)
-	if err != nil {
-		return nil, errors.Errorf("cannot parse application URL: %q", urlStr)
-	}
-	if url.RawQuery != "" || url.Fragment != "" || url.User != nil {
-		return nil, errors.Errorf("application URL %q has unrecognized parts", urlStr)
-	}
+	return parseApplicationURLParts(urlStr, true)
+}
 
+func parseApplicationURLParts(urlStr string, allowIncomplete bool) (*ApplicationURLParts, error) {
 	var result ApplicationURLParts
-	if url.Scheme != "" {
-		result.Directory = url.Scheme
-	}
-	urlPath := strings.Trim(url.Path, "/")
-	parts := strings.Split(urlPath, "/")
+	if modelApplicationRegexp.MatchString(urlStr) {
+		result.User = modelApplicationRegexp.ReplaceAllString(urlStr, "$user")
+		result.ModelName = modelApplicationRegexp.ReplaceAllString(urlStr, "$model")
+		result.ApplicationName = modelApplicationRegexp.ReplaceAllString(urlStr, "$application")
+	} else {
+		url, err := gourl.Parse(urlStr)
+		if err != nil {
+			return nil, errors.Errorf("cannot parse application URL: %q", urlStr)
+		}
+		if url.RawQuery != "" || url.Fragment != "" || url.User != nil {
+			return nil, errors.Errorf("application URL %q has unrecognized parts", urlStr)
+		}
 
-	if len(parts) < 2 {
-		switch len(parts) {
-		case 1:
-			result.ApplicationName = parts[0]
+		urlPath := strings.Trim(url.Path, "/")
+		if applicationURLRegexp.MatchString(urlPath) {
+			result.Directory = url.Scheme
+			result.User = applicationURLRegexp.ReplaceAllString(urlPath, "$user")
+			result.ApplicationName = applicationURLRegexp.ReplaceAllString(urlPath, "$application")
+			if result.User == "" && !allowIncomplete {
+				return nil, errors.Errorf("application URL has invalid form, missing %q: %q", "/u/<user>", urlStr)
+			}
+			if result.ApplicationName == "" && !allowIncomplete {
+				return nil, errors.Errorf("application URL has invalid form, missing application name: %q", urlStr)
+			}
+		} else {
+			return nil, errors.Errorf("application URL has invalid form: %q", urlStr)
 		}
-		return &result, nil
-	}
 
-	if parts[0] == "u" {
-		if !names.IsValidUser(parts[1]) {
-			return nil, errors.NotValidf("user name %q", parts[1])
+		// Validate the resulting URL part values.
```

Comment:
Should this validation for user/model/application be done for `model.application` as well, or is this covered completely by the regex matching?